### ISSUE TITLE
Awareness of compatibility packages

### ISSUE THREAD

**Issue by glynnforrest** [llaville/php-compatinfo#237]
Firstly, thank you for this great tool! I thought I was going to have to write a script to figure out what extensions a project required. I'm glad I can use this tool instead.

Do you think it would be possible for the script to detect usage of 'compat' packages, that bring newer features to older PHP versions?

For example, I have some code that uses `random_bytes()`, introduced in PHP 7. By requiring `paragonie/random_compat` I can have it work for older versions. It would be great if `php-compat-info` could detect this.

I guess this would require parsing the requirements in `composer.json` to see if any of these packages are being used.

Some package examples:

* https://github.com/symfony/polyfill
* https://github.com/paragonie/random_compat
* https://github.com/ircmaxell/password_compat

**Comment by llaville** [comment]
Many versions and years later, upcoming release 6.4 will finally include this new feature !

**Comment by llaville** [comment]
I'll push the code to a new branch **polyfill** tomorrow. Your feedback, and any test you could done will be welcome !

**Comment by llaville** [comment]
@glynnforrest Code is available.

We need to optimize process because polyfills initialization take too much time for me.

![compatinfo64-polyfills-warmup](https://user-images.githubusercontent.com/364342/157823589-8e817f47-8162-43e8-b1cb-489dea7f20ef.png)

Here is an example with PHP CompatInfo 6.4 code scanned

![compatinfo64-polyfills](https://user-images.githubusercontent.com/364342/157823688-853c354c-d085-425f-b610-29a82e05f70c.png)

Without polyfills, results are wrong and give PHP 8.0 as min requirements, while with polyfills, results are as expected.

You need to initialized
- as previous the database `db:create`, and `db:init`
- the polyfills depending of your source code installed : run `bin/polyfill-warmup`

And run data source code analyse

**Comment by llaville** [comment]
No need to optimize the `bin/polyfill-warmup` script. My Laptop was overloaded during tests ;-) It's run now as expected 

![polyfills-warmup](https://user-images.githubusercontent.com/364342/158006606-63c19dfc-8da8-4392-bb71-cc6586f044ac.png)

**Comment by llaville** [comment]
BTW, after a night to think about a better solution, I've in mind, in next days, to include polyfills handle to the https://github.com/llaville/php-compatinfo-db project, and adjust code in CompatInfo.

In the meantime, if somebody want to share is feedback about this feature before it will become official, he will be welcome !

**Comment by llaville** [comment]
Found origin of  double time elapsed with `bin/polyfill-warmup` script (and its uses in ReferenceCollection).

But even without this issue, init time is a bit long, and I'm +1 to include it in Compatinfo-DB now !

![polyfills-warmup](https://user-images.githubusercontent.com/364342/158009537-c865c5e6-114d-41eb-9139-cc0f775e2305.png)

Scheduled for PHP CompatInfoDB 4.2.0

**Comment by llaville** [comment]
CompatInfoDB 4.2.x-dev is ready to support polyfills. See commit https://github.com/llaville/php-compatinfo-db/commit/c5ce8ac4e85730c227193202b871ba261f6724ab

**Comment by llaville** [comment]
CompatInfo 6.4.x (with branch `polyfill`) and commit https://github.com/llaville/php-compatinfo/commit/9b3a7557a0335e596b0570bbe5f4051e4057f2c0 use now the CompatInfoDB 4.2.x-dev (branch `polyfill`) for detection.

NOTE: it won't be necessary in final version of CompatInfoDB 4.2.0 were symfony/polyfill will be added.

To add a polyfill (compatible) like https://github.com/symfony/polyfill/tree/main/src/Php80 do as follow : 

- run `db:create` 
- run `db:polyfill symfony/polyfill-php80 v1.25.0 --php 8.0 -v` 
- run `db:init`

**Comment by llaville** [comment]
Reminder: This report will be closed in 5 days even if I don't receive any feedback !

**Comment by glynnforrest** [comment]
Thanks for your work on this @llaville! I can't get this feature working, am I doing something wrong?

Here is an example of trying `symfony/polyfill-php80` on php 7.4.

```bash
php -v

# PHP 7.4.28 (cli) (built: Mar  1 2022 01:37:04) ( NTS )
```

Create `test.php` and try without polyfill:

```php
#!/usr/bin/env php
<?php

var_dump(str_starts_with('foobar', 'foo'));
```

```bash
./test.php

# PHP Fatal error:  Uncaught Error: Call to undefined function str_starts_with()
```

Add the polyfill:

```bash
composer require symfony/polfill-php80 ^1.25.0
```

```diff
  #!/usr/bin/env php
  <?php

+ include 'vendor/autoload.php';
  var_dump(str_starts_with('foobar', 'foo'));
```

```bash
./test.php

# bool(true)
```

Now analyse with `php-compatinfo`:

```bash
composer require bartlett/php-compatinfo ^6.4-dev

./vendor/bin/phpcompatinfo db:create
./vendor/bin/phpcompatinfo db:init # At least one distribution platform should exist. None available. Run "db:init" command to build one.
./vendor/bin/phpcompatinfo db:polyfill symfony/polyfill-php80 v1.25.0 --php 8.0 -v
./vendor/bin/phpcompatinfo db:create
```

I see polyfill analysis in the output, but still PHP 8 is required:

<img width="831" alt="image" src="https://user-images.githubusercontent.com/578458/161460114-afc3fd12-911f-49b5-8500-66e960e33a3a.png">

I tried with:

```
composer require bartlett/php-compatinfo '^6.4-dev'
composer require bartlett/php-compatinfo 'dev-polyfill'
```

**Comment by llaville** [comment]
@glynnforrest 
* **First**, I would like to thanks you to provide a such issue report with full details. It's so rare these days ;-)

* **Second**, I would like to clarify status code and procedure to install correctly PHP CompatInfo (6.4.0 upcoming version)

When you try to install source code version from a branch, take care to give the right constaint: 

With `composer require bartlett/php-compatinfo ^6.4-dev` you won't install polyfill code. Checks Composer output that give
```
- Installing bartlett/php-compatinfo-db (4.2.0): Extracting archive
- Installing bartlett/php-compatinfo (6.4.x-dev c19c87b): Extracting archive
```
Picking code from branch `6.4`

With following constraint
```
composer require bartlett/php-compatinfo dev-polyfill
```
You get the right code, verified by Composer output
```
  - Installing bartlett/php-compatinfo-db (4.2.0): Extracting archive
  - Installing bartlett/php-compatinfo (dev-polyfill 0f3664b): Extracting archive
```

Now, that Polyfills feature is supported by PHP CompatInfo DB 4.2.0, you don't have to specify the `db:polyfill` command. 
It's for me (maintainer of CompatInfoDB project) and user that want to add new polyfill in database not yet supported.

So, in summary : `db:create` and `db:init` is enough !

* **Third** you are right there is a little regression with available code, due to Unit tests. Fix will be push in few seconds

**Comment by llaville** [comment]
With commit 5f81d4d, unit tests and `analyser:run` command works as expected

![expected_results_issue237](https://user-images.githubusercontent.com/364342/161514697-ef03c65f-e080-46d9-b8c2-aab915ec83f6.png)

**Comment by llaville** [comment]
I've just opened a discussion https://github.com/llaville/php-compatinfo/discussions/350 to explain how to understand polyfill results
